src/ucs/sys/netlink.c

```diff
@@ -186,52 +212,108 @@ static ucs_status_t
 ucs_netlink_parse_rt_entry_cb(const struct nlmsghdr *nlh, const void *nl_msg,
                               void *arg)
 {
-    ucs_netlink_route_info_t *info = (ucs_netlink_route_info_t *)arg;
-    int rule_iface;
+    int iface_index;
     const void *dst_in_addr;
+    route_entry_t *new_rule;
+    ucs_list_link_t *iface_rules;
+    khiter_t iter;
+    int khret;
 
     if (ucs_netlink_get_route_info(RTM_RTA((const struct rtmsg *)nl_msg),
-                                   RTM_PAYLOAD(nlh), &rule_iface,
+                                   RTM_PAYLOAD(nlh), &iface_index,
                                    &dst_in_addr) != UCS_OK) {
         return UCS_INPROGRESS;
     }
 
-    if (rule_iface != info->if_index) {
-        return UCS_INPROGRESS;
+    iter = kh_put(ucs_netlink_rt_cache, &ucs_netlink_routing_table_cache,
+                  iface_index, &khret);
+    if (khret == UCS_KH_PUT_FAILED) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    /* get the pointer to the head of the rule list. If the iface was not
+       present in the hash table before, initialize the list */
+    iface_rules = &kh_val(&ucs_netlink_routing_table_cache, iter);
+    if (khret != UCS_KH_PUT_KEY_PRESENT) {
+        ucs_list_head_init(iface_rules);
     }
 
-    if (ucs_bitwise_is_equal(ucs_sockaddr_get_inet_addr(info->sa_remote),
-                             dst_in_addr,
-                             ((const struct rtmsg *)nl_msg)->rtm_dst_len)) {
-        info->found = 1;
-        return UCS_OK;
+    /* allocate and initialize a new rule element */
+    new_rule = ucs_malloc(sizeof(route_entry_t), "route entry");
+    if (new_rule == NULL) {
+        ucs_error("could not allocate route entry");
+        return UCS_ERR_NO_MEMORY;
     }
 
+    new_rule->dest.ss_family = ((const struct rtmsg *)nl_msg)->rtm_family;
+    if (((const struct rtmsg *)nl_msg)->rtm_family == AF_INET) {
+        memcpy(&((struct sockaddr_in *)&new_rule->dest)->sin_addr,
+               dst_in_addr, sizeof(struct in_addr));
+    } else {
+        memcpy(&((struct sockaddr_in6 *)&new_rule->dest)->sin6_addr,
+               dst_in_addr, sizeof(struct in6_addr));
+    }
+
+    new_rule->if_index    = iface_index;
+    new_rule->subnet_mask = ((const struct rtmsg *)nl_msg)->rtm_dst_len;
```

Comment:
Maybe rename to `subnet_prefix_len` or smth similar, to avoid confusion.
It is **not** a subnet mask, but number of bits.